### PR TITLE
Doc: Add warning to treat group chunks as immutable

### LINKED ISSUE #14180: DOC: clarify dangers of fast apply in GroupBy apply docs
I recently had to solve an issue where some code was behaving inconsistently. Effectively, I tracked this issue to a function that was being passed to apply which occasionally, under certain circumstances, mutated the input for a group - and not always the first group. Because fast_apply only checks the first group this meant that the output was unreliable. It didn't take me long to work out, but from a user's point of view it would be good to extend the docs so that they clarify that mutating the input is either not a good idea, or it should occur consistently for each group (or at the very least always for the first group).

### PR DESCRIPTION
**Pull request by pdpark**
- [X] closes #14180

### CODE REVIEW

**Comment by pdpark** [comment]
New clean pull request with only the changes for this issue.

File: doc/source/groupby.rst [hunk 1]
```diff
@@ -944,13 +944,17 @@ that is itself a series, and possibly upcast the result to a DataFrame:
    So depending on the path taken, and exactly what you are grouping. Thus the grouped columns(s) may be included in
    the output as well as set the indices.
 
-.. warning::
+.. warnings::
```
Comment (jreback) [inline doc/source/groupby.rst]: its warning

File: doc/source/groupby.rst [hunk 2]
```diff
@@ -944,13 +944,17 @@ that is itself a series, and possibly upcast the result to a DataFrame:
...
 
-    In the current implementation apply calls func twice on the
+    * In the current implementation apply calls func twice on the
     first group to decide whether it can take a fast or slow code
     path. This can lead to unexpected behavior if func has
     side-effects, as they will take effect twice for the first
     group.
+    
+    * Apply should not perform in-place operations on the group chunk.
```
Comment (jreback) [inline doc/source/groupby.rst]: apply functions

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by pdpark** [comment]
Requested changes made.

File: doc/source/groupby.rst [hunk 3]
```diff
@@ -946,11 +946,15 @@ that is itself a series, and possibly upcast the result to a DataFrame:
...
 .. warning::
 
-    In the current implementation apply calls func twice on the
+    * In the current implementation apply calls func twice on the
     first group to decide whether it can take a fast or slow code
     path. This can lead to unexpected behavior if func has
     side-effects, as they will take effect twice for the first
     group.
+    
+    * Apply should not apply functions to the group chunk.
```
Comment (jreback) [inline doc/source/groupby.rst]: this is not very clear here, its the *user* function that should not modify the group chunks.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@pandas-dev/pandas-core if someone wants to have a look here

File: doc/source/groupby.rst (hunk 3 above)
Comment (chris-b1) [reply doc/source/groupby.rst]: maybe instead
```
* for predictable results and performance, func should **not** mutate any of the group chunks
```

**Comment by jorisvandenbossche** [comment]
As discussed in one of the other issues (https://github.com/pandas-dev/pandas/issues/14927), I think we should rather deprecate the special cased behaviour when the group is mutated, than documenting it.

**Comment by toobaz** [comment]
> As discussed in one of the other issues (#14927), I think we should rather deprecate the special cased behaviour when the group is mutated, than documenting it.

If you mean "_the return value of ``apply()`` should not depend on whether the data was mutated, so the current special case is not worth documenting_", I agree.

But my understanding is that the actual changes to the original ``DataFrame`` are unpredictable/not guaranteed if you mutate the groups, and that this PR warns about that.

Did I misunderstand? If not, maybe the warning can be made more specific.

**Comment by jorisvandenbossche** [comment]
> But my understanding is that the actual changes to the original DataFrame are unpredictable/not guaranteed if you mutate the groups, and that this PR warns about that.

Ah, but I was not talking about changes to the *original* dataframe. Does that ever happen? (that would mean the group is a view on the original dataframe). I would like to see a code example for that. 

What I thought it was about that mutating the group (inplace), so the same *object* is returned from the function, leads to a different behaviour of how the results are back combined together. 
And it is this different behaviour that I would want to deprecate.

**Comment by toobaz** [comment]
> Ah, but I was not talking about changes to the original dataframe. Does that ever happen?

To be honest I have no idea. If it does not happen, we might then say (after fixing #14927 ) that mutating the chunks can be freely done, because/but it won't affect the original df.

(if it _can_ possibly happen, we might want to keep the warning, and make it refer specifically to the original df)

File: doc/source/groupby.rst (hunk 3 above)
Comment (pdpark) [reply doc/source/groupby.rst]: This has been resolved.

**Comment by pdpark** [comment]
How does this look?

**Comment by datapythonista** [comment]
@toobaz @jorisvandenbossche are you happy to merge the latest version of this, or should we close the PR instead?

**Pull request closed without merging**